mint/git-clone/mint-leaf.yml

```diff
@@ -77,13 +80,47 @@ tasks:
 
       git checkout ${{ params.ref }}
 
-      echo "Checked out git repository at $(git rev-parse HEAD)"
+      commit_sha=$(git rev-parse HEAD)
+      echo "Checked out git repository at ${commit_sha}"
 
       if [[ "${{ params.lfs }}" == "true" ]]; then
         git lfs fetch
         git lfs checkout
       fi
 
+      # Set metadata
+      printf "${{ params.repository }}" >> "$MINT_CACHE_KEY_EXCLUDED_ENV/MINT_GIT_REPOSITORY_URL"
+      printf "${{ params.repository }}" | tr ':' '/' | rev | cut -d '/' -f1,2 | rev | sed 's/\.git$//' >> "$MINT_CACHE_KEY_EXCLUDED_ENV/MINT_GIT_REPOSITORY_NAME"
+      commit_message=$(git log -n 1 --pretty=format:%B)
+      printf "${commit_message}" >> "$MINT_CACHE_KEY_EXCLUDED_ENV/MINT_GIT_COMMIT_MESSAGE"
+      printf "${commit_sha}" >> "$MINT_CACHE_KEY_EXCLUDED_ENV/MINT_GIT_COMMIT_SHA"
+
+      unresolved_ref=""
+      if [[ -n "${{ params.meta-ref }}" ]]; then
+        refs_matching_provided_ref=$(git for-each-ref --format="%(refname)" "refs/heads/${{ params.meta-ref }}" "refs/tags/${{ params.meta-ref }}" "${{ params.meta-ref }}" | grep -v refs/remotes)
+        unresolved_ref=$(echo "$refs_matching_provided_ref" | head -n 1 | tr -d '\n')
+
+        # also, ensure the meta-ref contains the resolved commit sha
+        result=$(git for-each-ref "${unresolved_ref}" --format="%(refname)" --contains "${commit_sha}")
+        if [[ -z "${result}" ]]; then
+          cat << EOF > $(mktemp "$MINT_ERRORS/error-XXXX")
+      The \`meta-ref\` provided does not contain the resolved commit sha.
+      EOF
+          exit 1
+        fi
+      elif [[ "${{ params.ref }}" == "${commit_sha}" ]]; then
+        refs_containing_sha=$(git for-each-ref --format="%(refname)" --contains | grep -v refs/remotes)
+        unresolved_ref=$(echo "$refs_containing_sha" | head -n 1 | tr -d '\n')
+      else
+        refs_matching_provided_ref=$(git for-each-ref --format="%(refname)" "refs/heads/${{ params.ref }}" "refs/tags/${{ params.ref }}" "${{ params.ref }}" | grep -v refs/remotes)
+        unresolved_ref=$(echo "$refs_matching_provided_ref" | head -n 1 | tr -d '\n')
+      fi
+
+      if [[ -n "${unresolved_ref}" ]]; then
+        printf "${unresolved_ref}" >> "$MINT_CACHE_KEY_EXCLUDED_ENV/MINT_GIT_REF"
+        printf "${unresolved_ref}" | sed -E 's|refs/[^/]+/||' >> "$MINT_CACHE_KEY_EXCLUDED_ENV/MINT_GIT_REF_NAME"
+      fi
+
       if [[ "${{ params.preserve-git-dir }}" == "false" ]]; then
         rm -rf .git
       fi
```

Comment:
In the case where we do preserve the git dir and where we infer the ref name by looking for a branch or tag with the commit, should we make sure that ref is checked out instead of just a detached commit?

Reply:
I'm not sure... I'm hesitant about all the edges around this. One pain point there, for example, is that checking out a tag always puts you in a detached head state

Reply:
Another edge in that area is that the commit might not be the most recent commit on that ref, so checking out the ref, then, would move you to a commit you might not expect

Reply:
This is why I had suggested (internally) that we may want to simply do a v2 of the leaf which requires `ref:` and expects it to be either a branch or tag name optionally prefixed by `refs/{heads,tags}` and never a commit sha. 

Then further optionally accepting a precise commit sha which must be contained within that ref

Reply:
We decided to punt on this for now offline